lib/rust/frp2/src/node/template.rs

```diff
@@ -0,0 +1,358 @@
+//! FRP node template implementation. It defines the node boilerplate logic for a given set of input
+//! types, allowing node definitions to be short and easy to read.
+
+use crate::prelude::*;
+use enso_generics::traits::*;
+
+use crate::data::Data;
+use crate::network::Network;
+use crate::node::class::NodeInNetwork;
+use crate::node::class::TypedNode;
+use crate::node::input;
+use crate::node::input::Listen;
+use crate::node::input::ListenAndSample;
+use crate::node::input::Sample;
+use crate::node::Node;
+use crate::node::NodeWithDefaultOutput;
+use crate::runtime::with_runtime;
+use crate::runtime::NodeData;
+use crate::runtime::Runtime;
+
+
+
+// ====================
+// === EventContext ===
+// ====================
+
+/// A wrapper for the runtime and current node data with explicit information about the event type.
+/// It is mainly used to make the API for defining nodes easy and type-safe.
+pub(crate) struct EventContext<'a, Output> {
+    tp:      PhantomData<Output>,
+    runtime: &'a Runtime,
+    node:    &'a NodeData,
+}
+
+impl<'a, Output> EventContext<'a, Output> {
+    /// Constructor. The `Output` type is not checked, so you need to be careful when using this
+    /// function.
+    #[inline(always)]
+    fn unchecked_new(runtime: &'a Runtime, node: &'a NodeData) -> Self {
+        EventContext { tp: PhantomData, runtime, node }
+    }
+}
+
+impl<'a, Output: Data> EventContext<'a, Output> {
+    /// Emit the output event.
+    #[inline(always)]
+    pub(crate) fn emit(self, value: &Output) {
+        self.runtime.unchecked_emit(self.node, value);
+    }
+}
+
+
+
+// ===================
+// === Model Usage ===
+// ===================
+
+/// Marker type indicating that the node uses the model.
+pub(crate) struct ModelUsed;
+
+/// Marker type indicating that the node does not use the model.
+pub(crate) struct ModelNotUsed;
+
+/// A phantom type passed to node body closure if the model was marked not to be used.
+pub(crate) struct ModelSkipped;
+
+/// Resolved model based on the model chooser and the FRP network model type.
+type ChosenModel<M, Model> = <M as ModelChooser<Model>>::ChosenModel;
+
+/// A trait allowing choosing either real FRP Network model or a phantom one based on the provided
+/// type-level configuration.
+trait ModelChooser<Model> {
+    type ChosenModel;
+    type ClonedModel: 'static;
+    fn clone_model(model: &Rc<OptRefCell<Model>>) -> Self::ClonedModel;
+    fn with_model_borrow_mut<Out>(
+        model: &Self::ClonedModel,
+        f: impl FnOnce(&mut Self::ChosenModel) -> Out,
+    ) -> Out;
```

Comment:
Will `model` be always _cloned_?

I admit I don't understand the method at this point; why it takes cloned model and pass _chosen_ model to the lambda?

Reply:
Yes, it is always cloned before using this function. See its usage below.